service/entityresolution/virtrusaas/v2/entity_resolution.go

```diff
@@ -0,0 +1,209 @@
+	ctx, span := s.Tracer.Start(ctx, "CreateEntityChainsFromTokens")
+	defer span.End()
+
+	resp, err := CreateEntityChainsFromTokens(ctx, req.Msg, s.logger)
+	return connect.NewResponse(&resp), err
+}
+
+func CreateEntityChainsFromTokens(
+	_ context.Context,
+	req *entityresolutionV2.CreateEntityChainsFromTokensRequest,
+	_ *logger.Logger,
+) (entityresolutionV2.CreateEntityChainsFromTokensResponse, error) {
+	entityChains := []*entity.EntityChain{}
+	resources := req.GetResources()
+
+	// for each token in the tokens form an entity chain
+	for _, tok := range req.GetTokens() {
+		entities, err := getEntitiesFromToken(tok.GetJwt(), resources)
+		if err != nil {
+			return entityresolutionV2.CreateEntityChainsFromTokensResponse{}, err
+		}
+		entityChains = append(entityChains, &entity.EntityChain{EphemeralId: tok.GetEphemeralId(), Entities: entities})
+	}
+
+	return entityresolutionV2.CreateEntityChainsFromTokensResponse{EntityChains: entityChains}, nil
+}
+
+func EntityResolution(_ context.Context,
+	req *entityresolutionV2.ResolveEntitiesRequest,
+	_ *logger.Logger,
+) (entityresolutionV2.ResolveEntitiesResponse, error) {
+	var resolvedEntities []*entityresolutionV2.EntityRepresentation
+
+	for idx, ident := range req.GetEntities() {
+		entityStruct := &structpb.Struct{}
+		switch ident.GetEntityType().(type) {
+		case *entity.Entity_Claims:
+			claims := ident.GetClaims()
+			if claims != nil {
+				err := claims.UnmarshalTo(entityStruct)
+				if err != nil {
+					return entityresolutionV2.ResolveEntitiesResponse{}, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("error unpacking anypb.Any to structpb.Struct: %w", err))
+				}
+			}
+			// make sure the id field is populated
+			originialID := ident.GetEphemeralId()
+			if originialID == "" {
+				originialID = ent.EntityIDPrefix + strconv.Itoa(idx)
+			}
+
+			directEntitlements := make([]*entityresolutionV2.DirectEntitlement, 0)
+			if pbstructDirectEntitlement, ok := entityStruct.GetFields()["direct_entitlements"]; ok {
+				for _, entitlement := range pbstructDirectEntitlement.GetListValue().GetValues() {
+					bytes, err := protojson.Marshal(entitlement)
+					if err != nil {
+						return entityresolutionV2.ResolveEntitiesResponse{}, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("error marshaling direct entitlement: %w", err))
+					}
+
+					var directEntitlement entityresolutionV2.DirectEntitlement
+					if err := protojson.Unmarshal(bytes, &directEntitlement); err != nil {
+						return entityresolutionV2.ResolveEntitiesResponse{}, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("error unmarshaling direct entitlement: %w", err))
+					}
+					directEntitlements = append(directEntitlements, &directEntitlement)
+				}
+
+				delete(entityStruct.GetFields(), "direct_entitlements")
```

Comment:
add comment as to why this is done (don't want subject mappings to process them)